added dvar to change next special round

lveez opened the pull request.

TTS4life commented:
I like having a dvar for special rounds. Is there any way we could implement this into a menu option to make it more easier for people to use? Like a special round toggle in options menu that constantly repeats special rounds for practice or something? (There's a bunch of examples for menu stuff in `ui/options_game.menu`

Just trying to brainstorm if there's any way we could make this more intuitive for people who might wanna practice special rounds. If not, Ill add some stuff in the readme for it noting its an option. Thanks!

lveez commented:
I added the option to choose your next dog/doc/monkey round (+1-4 from the start round), added a map option for a hudelem to track the average kills per waffe shot (not including dogs), and fixed the bug where sometimes you get loads of dogs spawning when you first load der riese.

Not sure if this is the best way to implement the menu for the dog rounds but the main use case I saw for it was to be able to practice several rounds (e.g 70-100) with dog rounds.

TTS4life commented on maps/_zombiemode.gsc:
```diff
@@ -4176,9 +4194,30 @@ round_think()
-	level.next_doc_round = 666;
+
+	// lveez - if don't wait for this flag the next doc rounds gets reset
+	if (level.script == "zombie_pentagon")
```
This section for handling special rounds at the being at the beginning of `round_think()` means that it is only going to get executed once when the game is setting things up for the first time when loading into a map or restarting level. I think this would do better a couple lines down inside of the infinite loop before the flag for "round_start". That way, people can choose on the fly how often they want the next special round to occur (e.g. I only want to practice monkey rounds on Ascension, I can do them back-to-back without having to wait the typical 4-5 rounds to practice them again). The only thing you'll need to look out for is that the original logic for the next special round does not overwrite the custom logic. Otherwise everything else is good.

lveez replied:
In the latest commit I changed the logic so that you can change the frequency of the dog rounds at any point, and gave options of: 4/5 rounders, 4 rounders, 5 rounders, every round, and off. I think this would cover all use cases.

I also fixed the bug properly since the old commit didn't actually fix it despite it working every test I did originally.

TTS4life reviewed (changes requested):
Looks good overall but I would like to see the tweaked a little so that I can set the next special round to be `X` rounds away whenever I please, instead of only once at the beginning of a game (see comment on `_zombiemode.gsc`). Once that change is made I'm looking forward to merging this for a new release. Thanks for the contribution!

TTS4life reviewed (changes requested):
Love the changes. Monkeys, Doc and Dogs can all now be set to occur every round. However, I'm experiencing a strange bug when trying to change the frequency mid-game.

If I switch the dog rounds from happening every single round to every 4/5 rounds (or any other option really), I seem to be getting an issue where the game struggles to spawn zombies, and I get a weird dog-round-but-not scenario. The round will start like a normal round, but dogs spawn in 2s and I hear "Fetch me their souls", as if its a dog round, despite not having the special riff that starts a special round, and the fog settling in. I also do not get a Max Ammo for killing them all. I did not see this issue with the doc or monkeys. I tested dogs rounds on both Kino and Der Riese and experienced this issue. Sometimes the round will skip in this state. I'm not sure what would be causing this issue offhand. I did not test the reverse, you might wanna check that too.

While you're investigating, could you change the "1" option in the "Special Round Frequency" to "Every Round"? I think it would read better than just 1 and match the "Never" option a little better. Otherwise I'm happy with this new menu option's behavior. Thanks for your hard work!

lveez commented:
Nice catch, only happened when 2 dog rounds were played in a row so didn't spot that one when I tested it. Should be all working now.
